### PR TITLE
feat(archive): Implement data retention. Closes #2273

### LINKED ISSUE #2273: Workflow Archive data retention
# Summary

Currently the workflow archive is append only and users need to manually clean-up data. It wight be good to have a way to automatically delete records.

# Motivation

The goals would be:

* Compliance - e.g. GDPR
* Cost - reduce database usage

# Proposal

A configuration item that sets how long to keep data for.

### PR DESCRIPTION
**Pull request by alexec**
Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [x] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlike to be merged.
* [x] Optional. I've added My organization is added to the README.
* [x] I've signed the CLA and required builds are green. 

Closes #2273

# Testing

This must be manually tested. 

Evidence of testing:

## Disabled

```
time="2020-04-17T20:16:58Z" level=info msg="Archived workflows TTL zero - so archived workflow GC disabled - you must restart the controller if you enable this"
```

## Enabled

```
time="2020-04-17T20:11:12Z" level=info msg="Performing archived workflow GC" periodicity=30s ttl=3m0s
...
time="2020-04-17T20:04:09Z" level=debug msg="Deleting offloaded nodes" uid=462323b6-3978-4fc3-b6ff-b19a79f9f565 version="fnv:525130960"
2020/04/17 20:04:09 
	Session ID:     00001
	Query:          DELETE FROM "argo_archived_workflows" WHERE ("clustername" = $1 AND finishedat < current_timestamp - interval '180' second)
	Arguments:      []interface {}{"default"}
	Rows affected:  2
	Time taken:     0.00305s
	Context:        context.Background

time="2020-04-17T20:04:09Z" level=info msg="Deleted archived workflows" rowsAffected=2
```

### CODE REVIEW

File: workflow/controller/controller.go [hunk 1]
```diff
@@ -349,6 +350,42 @@ func (wfc *WorkflowController) periodicWorkflowGarbageCollector(stopCh <-chan st
...
+		periodicity, err = time.ParseDuration(value)
+		if err != nil {
+			log.WithFields(log.Fields{"err": err, "value": value}).Fatal("Failed to parse ARCHIVED_WORKFLOW_GC_PERIOD")
+		}
+	}
+	if wfc.Config.Persistence == nil {
+		log.Info("Persistence disabled - so archived workflow GC disabled - you must restart the controller if you enable this")
+		return
+	}
+	ttl := wfc.Config.Persistence.ArchiveTTL
```
Comment (whynowy) [inline workflow/controller/controller.go:367]: `wfc.Config.Persistence==false` check is missing

File: persist/sqldb/workflow_archive.go [hunk 2]
```diff
@@ -211,3 +213,20 @@ func (r *workflowArchive) DeleteWorkflow(uid string) error {
 	log.WithFields(log.Fields{"uid": uid, "rowsAffected": rowsAffected}).Debug("Deleted archived workflow")
 	return nil
 }
+
+func (r *workflowArchive) DeleteWorkflows(ttl time.Duration) error {
+	rs, err := r.session.
+		DeleteFrom(archiveTableName).
+		Where(db.Cond{"clustername": r.clusterName}).
```
[the remark this answers is not in the document]
Comment (alexec) [reply persist/sqldb/workflow_archive.go:220]: ek - you're correct

**Comment by alexec** [comment]
@whynowy changes made as requested. Please can you re-review?

**Comment by alexec** [comment]
Re-run from failed.

**Comment by alexec** [comment]
Rerun from failed.

**Pull request merged**